.github/utils/generate_json_schema.py

```diff
@@ -0,0 +1,256 @@
+from typing import Any, Dict, Optional, Set, Tuple
+
+import haystack.document_stores
+import haystack.nodes
+import pydantic.schema
+from fastapi.dependencies.utils import get_typed_signature
+from github import Github
+from pydantic import BaseConfig, BaseSettings, Required, SecretStr, create_model
+from pydantic.fields import ModelField
+from pydantic.schema import SkipField, TypeModelOrEnum, TypeModelSet, encode_default
+from pydantic.schema import field_singleton_schema as _field_singleton_schema
+from pydantic.typing import is_callable_type
+from pydantic.utils import lenient_issubclass
+
+schema_version = "0.7"
+filename = f"haystack-pipeline.{schema_version}.schema.json"
+destination_path = Path(__file__).parent.parent.parent / "json-schemas" / filename
+
+
+class Settings(BaseSettings):
+    input_token: SecretStr
+    github_repository: str
+
+
+# Monkey patch Pydantic's field_singleton_schema to convert classes and functions to
+# strings in JSON Schema
+def field_singleton_schema(
+    field: ModelField,
+    *,
+    by_alias: bool,
+    model_name_map: Dict[TypeModelOrEnum, str],
+    ref_template: str,
+    schema_overrides: bool = False,
+    ref_prefix: Optional[str] = None,
+    known_models: TypeModelSet,
+) -> Tuple[Dict[str, Any], Dict[str, Any], Set[str]]:
+    try:
+        return _field_singleton_schema(
+            field,
+            by_alias=by_alias,
+            model_name_map=model_name_map,
+            ref_template=ref_template,
+            schema_overrides=schema_overrides,
+            ref_prefix=ref_prefix,
+            known_models=known_models,
+        )
+    except (ValueError, SkipField):
+        schema: Dict[str, Any] = {"type": "string"}
+
+        if isinstance(field.default, type) or is_callable_type(field.default):
+            default = field.default.__name__
+        else:
+            default = field.default
+        if not field.required:
+            schema["default"] = encode_default(default)
+        return schema, {}, set()
+
+
+# Monkeypatch Pydantic's field_singleton_schema
+pydantic.schema.field_singleton_schema = field_singleton_schema
+
+
+class Config(BaseConfig):
+    extra = "forbid"
+
+
+def get_json_schema():
+    schema_definitions = {}
+    additional_definitions = {}
+
+    modules_with_nodes = [haystack.nodes, haystack.document_stores]
+    possible_nodes = []
+    for module in modules_with_nodes:
+        for importable_name in dir(module):
+            imported = getattr(module, importable_name)
+            possible_nodes.append((module, imported))
+    # TODO: decide if there's a better way to not include Base classes other than by
+    # the prefix "Base" in the name. Maybe it could make sense to have a list of
+    # all the valid nodes to include in the main source code and then using that here.
+    for module, node in possible_nodes:
+        if lenient_issubclass(
+            node, haystack.nodes.BaseComponent
+        ) and not node.__name__.startswith("Base"):
```

Comment:
Note for a future refactoring: we should probably make our base classes subclasses of `abc.ABC` (https://docs.python.org/3/library/abc.html) to properly highlight the fact that they should not be used directly.

This should not be part of this PR, however I can imagine that the implementation could look like:
```python
if lenient_issubclass(node, BaseComponent) and not issubclass(node, abc.ABC):
```